### DISCUSSION TITLE
Slim down java project to ReadExample

### DISCUSSION THREAD

**Discussion by daddebackstrom** [eclipse-milo/milo#802] [General]
Hello Milo, and thank you so much for letting us use your project to connect to OPC servers :)

I want to use your project to connect to and read ONE value from a Kepware OPC Server Ex V6 instance. I suppose it's the ReadExample class that could be used as a template here?
 
If I want to extract that example to my own slim project, which jar-files would I need? I obviously don't need the server part...

Regards,
David

**Comment by kevinherron** [comment]
Hopefully you're using some kind of build tool like Maven or Gradle that resolves dependencies for you.

The main artifact you'll need is `sdk-client.jar` (and all of its dependencies). The Maven dependency would look something like this:
```
<dependency>
    <groupId>org.eclipse.milo</groupId>
    <artifactId>sdk-client</artifactId>
    <version>0.5.4</version>
</dependency>
```

Use the examples as a guide, not a template. I see people trying to copy the structure of ClientExampleRunner into their application and that's not really the point; it just exists to run examples, not as something you should be copying. The stuff to pay attention to is in the example itself.

**Reply by daddebackstrom** [reply]
Hello Kevin, and thank you for a quick reply :)

Yes I'm using Maven and I did see the sdk-client artifact. I just wasn't sure that the underlying sdk's were incorporated in that artifact.

I'll use the examples with care :) But maybe I'm looking in the wrong place, I was looking for an example were you connect to a server and specify a specific Tag to read from. Do you have such an example, or it there a tutorial availabe?

Regards,
David

**Reply by kevinherron** [reply]
No it's fine, this is the right example. I'm just saying don't copy every single part of it verbatim.

**Reply by daddebackstrom** [reply]
Hello again Kevin!

I've come quite far and managed to create a little spring boot application with a REST endpoint where another client is able to call my client and ask for a value (see the code below). My question now is: is this best practice? Doing read after read? Should I have a subscription running instead? I don't want to overload the KepWare server....

Again, thank you so much for taking the time so answer a newbie :)

@RequestMapping("/readnodevalue")
    public NodeValue readNodeValue(@RequestParam(value = "nodeId", defaultValue = "") String nodeId) {

        try {
            DataValue dataValue = opcUaClient.readValue(0.0, TimestampsToReturn.Both, new NodeId(2, nodeId)).get();

            String value = dataValue.getValue().getValue().toString();

            logger.info("Identifier={}, Value={}", nodeId, value);

            return new NodeValue(nodeId, value);

        } catch (Exception e) {
            logger.info("Error! - {}", e.getMessage());
            return new NodeValue(nodeId, "0.0");
        }
    }

**Reply by kevinherron** [reply]
I think it probably depends how often the endpoint will be called and if it's always going to be for the same NodeId. 

A subscription is generally more efficient but with that you will have more bookkeeping and setup/teardown to do.

**Reply by daddebackstrom** [reply]
The value will be read at maximum once an hour and right now it will always be the same node.

**Reply by kevinherron** [reply]
Using a read instead of subscribing is completely fine then.

**Reply by daddebackstrom** [reply]
Ok, thank you very much Kevin! Your stack works really well, nice job :) Take care
